src/statistics.rs

```diff
@@ -0,0 +1,771 @@
+        } else {
+            general_requirements += " AND chat_id NOT IN temp.one_one_chats";
+        }
+        let params = (ContactId::SELF, statistics_bot_chat_id, last_counted_msg);
+
+        let to_verified = t.query_row(
+            &format!(
+                "SELECT COUNT(*) FROM msgs
+                WHERE chat_id IN temp.verified_chats
+                AND {general_requirements}"
+            ),
+            params,
+            |row| row.get(0),
+        )?;
+
+        let unverified_encrypted = t.query_row(
+            &format!(
+                // (param GLOB '*\nc=1*' OR param GLOB 'c=1*') matches all messages that are end-to-end encrypted
+                "SELECT COUNT(*) FROM msgs
+                WHERE chat_id NOT IN temp.verified_chats AND chat_id NOT IN temp.empty_chats
+                AND (param GLOB '*\nc=1*' OR param GLOB 'c=1*')
+                AND {general_requirements}"
+            ),
+            params,
+            |row| row.get(0),
+        )?;
+
+        let unencrypted = t.query_row(
+            &format!(
+                "SELECT COUNT(*) FROM msgs
+                WHERE chat_id NOT IN temp.verified_chats AND chat_id NOT IN temp.empty_chats
+                AND NOT (param GLOB '*\nc=1*' OR param GLOB 'c=1*')
+                AND {general_requirements}"
+            ),
+            params,
+            |row| row.get(0),
+        )?;
+
+        let only_to_self = t.query_row(
+            &format!(
+                "SELECT COUNT(*) FROM msgs
+                WHERE chat_id IN temp.empty_chats
+                AND {general_requirements}"
+            ),
+            params,
+            |row| row.get(0),
+        )?;
+
+        t.execute("DROP TABLE temp.verified_chats", ())?;
+        t.execute("DROP TABLE temp.empty_chats", ())?;
+        t.execute("DROP TABLE temp.one_one_chats", ())?;
+
+        Ok(MessageStats {
+            to_verified,
+            unverified_encrypted,
+            unencrypted,
+            only_to_self,
+        })
+    };
+
+    let query_only = true;
+    let message_stats: MessageStats = context.sql.transaction_ex(query_only, trans_fn).await?;
+
+    Ok(message_stats)
+}
+
+pub(crate) async fn count_securejoin_ux_info(
+    context: &Context,
+    source: Option<u32>,
+    uipath: Option<u32>,
+) -> Result<()> {
+    if !should_send_statistics(context).await? {
+        return Ok(());
+    }
+
+    let source = source
+        .context("Missing securejoin source")
+        .log_err(context)
+        .unwrap_or(0);
+
+    context
+        .sql
+        .execute(
+            "INSERT INTO statistics_securejoin_sources VALUES (?, 1)
+                ON CONFLICT (source) DO UPDATE SET count=count+1;",
+            (source,),
+        )
+        .await?;
+
+    // We only get a UI path if the source is a QR code scan,
+    // a loaded image, or a link pasted from the QR code,
+    // so, no need to log an error if `uipath` is None:
+    let uipath = uipath.unwrap_or(0);
+    context
```

Comment:
Probably updating both tables in a single transaction isn't much more code, but guarantees that the counters never diverge